// @filoz/synapse-core/dist/src/piece/resolve-piece-url.js
import { request } from 'iso-web/http';
import pLocate from 'p-locate';
import pSome from 'p-some';
import { asChain } from "../chains.js";
import { findPiece } from "../sp/sp.js";
import { createPieceUrlPDP } from "../utils/piece-url.js";
import { getPdpDataSets } from "../warm-storage/get-pdp-data-sets.js";
export const defaultResolvers = [filbeamResolver, chainResolver];
export async function resolvePieceUrl(options) {
    const { address, client, pieceCid, signal, resolvers = defaultResolvers } = options;
    asChain(client.chain);
    const controller = new AbortController();
    const _signal = signal ? AbortSignal.any([controller.signal, signal]) : controller.signal;
    const result = await pSome(resolvers.map((resolver) => resolver({ address, client, pieceCid, signal: _signal })), { count: 1 });
    controller.abort();
    return result[0];
}
export async function filbeamResolver(options) {
    const { address, pieceCid, signal } = options;
    const chain = asChain(options.client.chain);
    if (chain.filbeam == null) {
        throw new Error('FilBeam not supported on this chain');
    }
    const url = `https://${address}.${chain.filbeam.retrievalDomain}/${pieceCid.toString()}`;
    const result = await request.head(url, {
        signal,
    });
    if (result.error) {
        throw result.error;
    }
    return url;
}
export async function chainResolver(options) {
    const { address, client, pieceCid, signal } = options;
    const dataSets = await getPdpDataSets(client, {
        address,
    });
    const providersById = dataSets.reduce((acc, dataSet) => {
        if (dataSet.live && dataSet.managed && dataSet.pdpEndEpoch === 0n) {
            acc.set(dataSet.providerId, dataSet.provider);
        }
        return acc;
    }, new Map());
    const providers = [...providersById.values()];
    const result = await findPieceOnProviders(providers, pieceCid, signal);
    if (result == null) {
        throw new Error('No provider found');
    }
    return createPieceUrlPDP({
        cid: pieceCid.toString(),
        serviceURL: result.pdp.serviceURL,
    });
}
export function providersResolver(providers) {
    return async (options) => {
        const { pieceCid, signal } = options;
        const result = await findPieceOnProviders(providers, pieceCid, signal);
        if (result == null) {
            throw new Error('No provider found');
        }
        return createPieceUrlPDP({
            cid: pieceCid.toString(),
            serviceURL: result.pdp.serviceURL,
        });
    };
}
export async function findPieceOnProviders(providers, pieceCid, signal) {
    const controller = new AbortController();
    const _signal = signal ? AbortSignal.any([controller.signal, signal]) : controller.signal;
    const result = await pLocate(providers.map((p) => findPiece({
        serviceURL: p.pdp.serviceURL,
        pieceCid,
        signal: _signal,
    }).then(() => p, () => null)), (p) => {
        if (p !== null) {
            controller.abort();
            return true;
        }
        return false;
    }, { concurrency: 5 });
    return result;
}
