// @filoz/synapse-core/src/session-key/authorization-expiry.ts
import type { Simplify } from 'type-fest'
import {
  type Address,
  type Chain,
  type Client,
  ContractFunctionExecutionError,
  type ContractFunctionParameters,
  type ContractFunctionReturnType,
  createClient,
  custom,
  type MulticallErrorType,
  type ReadContractErrorType,
  type Transport,
} from 'viem'
import { multicall, readContract } from 'viem/actions'
import type { sessionKeyRegistry as sessionKeyRegistryAbi } from '../abis/index.ts'
import { asChain } from '../chains.ts'
import type { ActionCallChain } from '../types.ts'
import { DefaultFwssPermissions, type Expirations, type Permission } from './permissions.ts'

export namespace authorizationExpiry {
  export type OptionsType = {
    /** The address of the user account. */
    address: Address
    /** The address of the session key. */
    sessionKeyAddress: Address
    /** The session key permission. */
    permission: Permission
    /** Session key registry contract address. If not provided, the default is the session key registry contract address for the chain. */
    contractAddress?: Address
  }

  export type ContractOutputType = ContractFunctionReturnType<
    typeof sessionKeyRegistryAbi,
    'pure' | 'view',
    'authorizationExpiry'
  >

  /** The expiry timestamp as a bigint (Unix timestamp in seconds). */
  export type OutputType = bigint

  export type ErrorType = asChain.ErrorType | ReadContractErrorType
}

/**
 * Get the authorization expiry timestamp for a session key permission.
 *
 * Returns the Unix timestamp (in seconds) when the authorization for the given
 * address, sessionKeyAddress, and permission combination expires. Returns 0 if no authorization exists.
 *
 * @param client - The client to use to get the authorization expiry.
 * @param options - {@link authorizationExpiry.OptionsType}
 * @returns The expiry timestamp as a bigint (Unix timestamp in seconds) {@link authorizationExpiry.OutputType}
 * @throws Errors {@link authorizationExpiry.ErrorType}
 *
 * @example
 * ```ts
 * import { authorizationExpiry, CreateDataSetPermission } from '@filoz/synapse-core/session-key'
 * import { createPublicClient, http } from 'viem'
 * import { calibration } from '@filoz/synapse-core/chains'
 *
 * const client = createPublicClient({
 *   chain: calibration,
 *   transport: http(),
 * })
 *
 * const expiry = await authorizationExpiry(client, {
 *   address: '0x1234567890123456789012345678901234567890',
 *   sessionKeyAddress: '0xabcdefabcdefabcdefabcdefabcdefabcdefabcd',
 *   permission: CreateDataSetPermission,
 * })
 *
 * console.log('Authorization expires at:', expiry)
 * ```
 */
export async function authorizationExpiry(
  client: Client<Transport, Chain>,
  options: authorizationExpiry.OptionsType
): Promise<authorizationExpiry.OutputType> {
  const data = await readContract(
    client,
    authorizationExpiryCall({
      chain: client.chain,
      address: options.address,
      sessionKeyAddress: options.sessionKeyAddress,
      permission: options.permission,
      contractAddress: options.contractAddress,
    })
  )
  return data
}

export namespace authorizationExpiryCall {
  export type OptionsType = Simplify<authorizationExpiry.OptionsType & ActionCallChain>
  export type ErrorType = asChain.ErrorType
  export type OutputType = ContractFunctionParameters<
    typeof sessionKeyRegistryAbi,
    'pure' | 'view',
    'authorizationExpiry'
  >
}

/**
 * Create a call to the authorizationExpiry function
 *
 * This function is used to create a call to the authorizationExpiry function for use with the multicall or readContract function.
 *
 * @param options - {@link authorizationExpiryCall.OptionsType}
 * @returns The call to the authorizationExpiry function {@link authorizationExpiryCall.OutputType}
 * @throws Errors {@link authorizationExpiryCall.ErrorType}
 *
 * @example
 * ```ts
 * import { authorizationExpiryCall, CreateDataSetPermission } from '@filoz/synapse-core/session-key'
 * import { createPublicClient, http } from 'viem'
 * import { multicall } from 'viem/actions'
 * import { calibration } from '@filoz/synapse-core/chains'
 *
 * const client = createPublicClient({
 *   chain: calibration,
 *   transport: http(),
 * })
 *
 * const results = await multicall(client, {
 *   contracts: [
 *     authorizationExpiryCall({
 *       chain: calibration,
 *       address: '0x1234567890123456789012345678901234567890',
 *       sessionKeyAddress: '0xabcdefabcdefabcdefabcdefabcdefabcdefabcd',
 *       permission: CreateDataSetPermission,
 *     }),
 *   ],
 * })
 *
 * console.log(results[0])
 * ```
 */
export function authorizationExpiryCall(options: authorizationExpiryCall.OptionsType) {
  const chain = asChain(options.chain)
  return {
    abi: chain.contracts.sessionKeyRegistry.abi,
    address: options.contractAddress ?? chain.contracts.sessionKeyRegistry.address,
    functionName: 'authorizationExpiry',
    args: [options.address, options.sessionKeyAddress, options.permission],
  } satisfies authorizationExpiryCall.OutputType
}

export namespace isExpired {
  export type OptionsType = Simplify<authorizationExpiry.OptionsType>
  export type ErrorType = authorizationExpiry.ErrorType
  export type OutputType = boolean
}

/**
 * Check if the session key is expired.
 *
 * @param client - The client to use.
 * @param options - The options to use.
 * @returns Whether the session key is expired.
 * @throws - {@link isExpired.ErrorType} if the read contract fails.
 */
export async function isExpired(
  client: Client<Transport, Chain>,
  options: isExpired.OptionsType
): Promise<isExpired.OutputType> {
  const expiry = await authorizationExpiry(client, options)

  return expiry < BigInt(Math.floor(Date.now() / 1000))
}

export namespace getExpirations {
  export type OptionsType = Simplify<
    Omit<authorizationExpiry.OptionsType, 'permission'> & { permissions?: Permission[] }
  >
  export type ErrorType = authorizationExpiry.ErrorType | MulticallErrorType
  export type OutputType = Record<Permission, bigint>
}

/**
 * Get the expirations for all FWSS permissions.
 *
 * @param client - The client to use.
 * @param options - {@link getExpirations.OptionsType}
 * @returns Expirations {@link getExpirations.OutputType}
 * @throws Errors {@link getExpirations.ErrorType}
 *
 * @example
 * ```ts
 * import { getExpirations } from '@filoz/synapse-core/session-key'
 * import { createPublicClient, http } from 'viem'
 * import { calibration } from '@filoz/synapse-core/chains'
 *
 * const client = createPublicClient({
 *   chain: calibration,
 *   transport: http(),
 * })
 *
 * const expirations = await getExpirations(client, {
 *   address: '0x1234567890123456789012345678901234567890',
 *   sessionKeyAddress: '0xabcdefabcdefabcdefabcdefabcdefabcdefabcd',
 * })
 *
 * console.log(expirations)
 */
export async function getExpirations(client: Client<Transport, Chain>, options: getExpirations.OptionsType) {
  const permissions = options.permissions ?? DefaultFwssPermissions
  const expirations: Expirations = Object.fromEntries(permissions.map((permission) => [permission, 0n]))

  // Use a plain client without an account for read-only calls. When a client
  // has an account, viem includes it as `from` in eth_call. On Filecoin, if
  // that address is not a deployed actor (common for session keys that have
  // never held FIL), Lotus rejects the call with "actor not found".
  const readClient = createClient({
    chain: client.chain,
    transport: custom({ request: client.transport.request }),
  })

  try {
    const result = await multicall(readClient, {
      allowFailure: false,
      contracts: permissions.map((permission) =>
        authorizationExpiryCall({
          chain: client.chain,
          address: options.address,
          sessionKeyAddress: options.sessionKeyAddress,
          permission,
        })
      ),
    })

    for (let i = 0; i < permissions.length; i++) {
      expirations[permissions[i]] = result[i]
    }
  } catch (e) {
    if (!(e instanceof ContractFunctionExecutionError && e.details?.includes('actor not found') === true)) {
      throw e
    }
  }

  return expirations
}
