// @filoz/synapse-core/dist/src/session-key/authorization-expiry.js
import { ContractFunctionExecutionError, createClient, custom, } from 'viem';
import { multicall, readContract } from 'viem/actions';
import { asChain } from "../chains.js";
import { DefaultFwssPermissions } from "./permissions.js";
export async function authorizationExpiry(client, options) {
    const data = await readContract(client, authorizationExpiryCall({
        chain: client.chain,
        address: options.address,
        sessionKeyAddress: options.sessionKeyAddress,
        permission: options.permission,
        contractAddress: options.contractAddress,
    }));
    return data;
}
export function authorizationExpiryCall(options) {
    const chain = asChain(options.chain);
    return {
        abi: chain.contracts.sessionKeyRegistry.abi,
        address: options.contractAddress ?? chain.contracts.sessionKeyRegistry.address,
        functionName: 'authorizationExpiry',
        args: [options.address, options.sessionKeyAddress, options.permission],
    };
}
export async function isExpired(client, options) {
    const expiry = await authorizationExpiry(client, options);
    return expiry < BigInt(Math.floor(Date.now() / 1000));
}
export async function getExpirations(client, options) {
    const permissions = options.permissions ?? DefaultFwssPermissions;
    const expirations = Object.fromEntries(permissions.map((permission) => [permission, 0n]));
    const readClient = createClient({
        chain: client.chain,
        transport: custom({ request: client.transport.request }),
    });
    try {
        const result = await multicall(readClient, {
            allowFailure: false,
            contracts: permissions.map((permission) => authorizationExpiryCall({
                chain: client.chain,
                address: options.address,
                sessionKeyAddress: options.sessionKeyAddress,
                permission,
            })),
        });
        for (let i = 0; i < permissions.length; i++) {
            expirations[permissions[i]] = result[i];
        }
    }
    catch (e) {
        if (!(e instanceof ContractFunctionExecutionError && e.details?.includes('actor not found') === true)) {
            throw e;
        }
    }
    return expirations;
}
