// @filoz/synapse-core/dist/src/session-key/login.js
import { parseEventLogs } from 'viem';
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import * as Abis from "../abis/index.js";
import { asChain } from "../chains.js";
import { DefaultFwssPermissions } from "./permissions.js";
export async function login(client, options) {
    const { request } = await simulateContract(client, loginCall({
        chain: client.chain,
        address: options.address,
        permissions: options.permissions,
        expiresAt: options.expiresAt,
        origin: options.origin,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function loginSync(client, options) {
    const hash = await login(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractLoginEvent(receipt.logs);
    return { receipt, event };
}
export function loginCall(options) {
    const chain = asChain(options.chain);
    const expiresAt = BigInt(Math.floor(Date.now() / 1000) + 3600);
    const permissions = options.permissions ?? DefaultFwssPermissions;
    return {
        abi: chain.contracts.sessionKeyRegistry.abi,
        address: options.contractAddress ?? chain.contracts.sessionKeyRegistry.address,
        functionName: 'login',
        args: [
            options.address,
            options.expiresAt ?? expiresAt,
            Array.from(new Set(permissions)),
            options.origin ?? 'synapse',
        ],
    };
}
export function extractLoginEvent(logs) {
    const [log] = parseEventLogs({
        abi: Abis.sessionKeyRegistry,
        logs,
        eventName: 'AuthorizationsUpdated',
        strict: true,
    });
    if (!log)
        throw new Error('`AuthorizationsUpdated` event not found.');
    return log;
}
