// @filoz/synapse-core/dist/src/session-key/permissions.js
import { TypedData } from 'ox';
import { keccak256, stringToHex } from 'viem';
import { EIP712Types } from "../typed-data/type-definitions.js";
function typeHash(type) {
    return keccak256(stringToHex(TypedData.encodeType(type)));
}
export const CreateDataSetPermission = typeHash({
    types: EIP712Types,
    primaryType: 'CreateDataSet',
});
export const AddPiecesPermission = typeHash({
    types: EIP712Types,
    primaryType: 'AddPieces',
});
export const SchedulePieceRemovalsPermission = typeHash({
    types: EIP712Types,
    primaryType: 'SchedulePieceRemovals',
});
export const DeleteDataSetPermission = typeHash({
    types: EIP712Types,
    primaryType: 'DeleteDataSet',
});
export const DefaultFwssPermissions = [
    CreateDataSetPermission,
    AddPiecesPermission,
    SchedulePieceRemovalsPermission,
    DeleteDataSetPermission,
];
export const DefaultEmptyExpirations = {
    [CreateDataSetPermission]: 0n,
    [AddPiecesPermission]: 0n,
    [SchedulePieceRemovalsPermission]: 0n,
    [DeleteDataSetPermission]: 0n,
};
