// @filoz/synapse-core/dist/src/session-key/revoke.js
import { simulateContract, waitForTransactionReceipt, writeContract } from 'viem/actions';
import { asChain } from "../chains.js";
import { extractLoginEvent } from "./login.js";
import { DefaultFwssPermissions } from "./permissions.js";
export async function revoke(client, options) {
    const { request } = await simulateContract(client, revokeCall({
        chain: client.chain,
        address: options.address,
        permissions: options.permissions,
        origin: options.origin,
        contractAddress: options.contractAddress,
    }));
    return writeContract(client, request);
}
export async function revokeSync(client, options) {
    const hash = await revoke(client, options);
    if (options.onHash) {
        options.onHash(hash);
    }
    const receipt = await waitForTransactionReceipt(client, { hash });
    const event = extractRevokeEvent(receipt.logs);
    return { receipt, event };
}
export function revokeCall(options) {
    const chain = asChain(options.chain);
    const permissions = options.permissions ?? DefaultFwssPermissions;
    return {
        abi: chain.contracts.sessionKeyRegistry.abi,
        address: options.contractAddress ?? chain.contracts.sessionKeyRegistry.address,
        functionName: 'revoke',
        args: [options.address, Array.from(new Set(permissions)), options.origin ?? 'synapse'],
    };
}
export const extractRevokeEvent = extractLoginEvent;
