// @filoz/synapse-core/dist/src/session-key/secp256k1.js
import { TypedEventTarget } from 'iso-web/event-target';
import { createClient, http, } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { watchContractEvent } from 'viem/actions';
import { asChain } from "../chains.js";
import { getExpirations } from "./authorization-expiry.js";
import { extractLoginEvent } from "./login.js";
import { DefaultEmptyExpirations } from "./permissions.js";
class Secp256k1SessionKey extends TypedEventTarget {
    #client;
    #type;
    #expirations;
    #unsubscribe;
    constructor(options) {
        super();
        this.#type = 'Secp256k1';
        this.#expirations = options.expirations;
        this.#client = options.client;
    }
    get client() {
        return this.#client;
    }
    get type() {
        return this.#type;
    }
    get expirations() {
        return this.#expirations;
    }
    get address() {
        return this.#client.account.address;
    }
    get rootAddress() {
        return this.#client.account.rootAddress;
    }
    get account() {
        return this.#client.account;
    }
    async watch() {
        await this.syncExpirations();
        if (!this.#unsubscribe) {
            this.#unsubscribe = watchContractEvent(this.client, {
                address: this.#client.chain.contracts.sessionKeyRegistry.address,
                abi: this.#client.chain.contracts.sessionKeyRegistry.abi,
                eventName: 'AuthorizationsUpdated',
                args: { identity: this.#client.account.rootAddress },
                onError: this.emit.bind(this, 'error'),
                onLogs: (logs) => {
                    try {
                        const event = extractLoginEvent(logs);
                        if (event.args.identity === this.#client.account.rootAddress) {
                            for (const hash of event.args.permissions) {
                                this.#expirations[hash] = event.args.expiry;
                            }
                            this.emit('expirationsUpdated', this.#expirations);
                        }
                    }
                    catch (error) {
                        this.emit('error', error);
                    }
                },
            });
            this.emit('connected', this.#expirations);
        }
        return () => {
            this.unwatch();
        };
    }
    unwatch() {
        if (this.#unsubscribe) {
            this.#unsubscribe();
            this.#unsubscribe = undefined;
            this.emit('disconnected');
        }
    }
    hasPermission(permission) {
        return this.expirations[permission] > BigInt(Math.floor(Date.now() / 1000));
    }
    hasPermissions(permissions) {
        return permissions.every((permission) => this.hasPermission(permission));
    }
    async syncExpirations(permissions) {
        this.#expirations = await getExpirations(this.#client, {
            address: this.#client.account.rootAddress,
            sessionKeyAddress: this.#client.account.address,
            permissions: permissions,
        });
        this.emit('expirationsUpdated', this.#expirations);
    }
}
export function fromSecp256k1(options) {
    const rootAddress = typeof options.root === 'string' ? options.root : options.root.address;
    if (rootAddress === undefined) {
        throw new Error('Root address is required');
    }
    const account = accountFromSecp256k1({
        privateKey: options.privateKey,
        rootAddress: rootAddress,
    });
    const chain = asChain(options.chain);
    const client = createClient({
        chain: chain,
        transport: options.transport ?? http(),
        account,
        name: 'Secp256k1 Session Key',
        key: 'secp256k1-session-key',
        type: 'sessionClient',
    });
    return new Secp256k1SessionKey({
        client: client,
        expirations: options.expirations ?? DefaultEmptyExpirations,
    });
}
export function accountFromSecp256k1(options) {
    const account = {
        ...privateKeyToAccount(options.privateKey),
        source: 'sessionKey',
        keyType: 'Secp256k1',
        rootAddress: options.rootAddress,
    };
    return account;
}
