// @filoz/synapse-core/dist/src/devnet/index.js
import * as Abis from "../abis/index.js";
export function toChain(devnetInfo) {
    const { info } = devnetInfo;
    const contracts = info.contracts;
    return {
        id: 31415926,
        name: 'FOC DevNet',
        nativeCurrency: {
            decimals: 18,
            name: 'Filecoin',
            symbol: 'FIL',
        },
        rpcUrls: {
            default: { http: [info.lotus.host_rpc_url] },
            public: { http: [info.lotus.host_rpc_url] },
        },
        blockExplorers: {
            default: {
                name: 'DevNet',
                url: 'http://localhost:3000',
            },
        },
        contracts: {
            multicall3: {
                address: contracts.multicall3_addr,
                blockCreated: 0,
            },
            usdfc: {
                address: contracts.mockusdfc_addr,
                abi: Abis.erc20WithPermit,
            },
            filecoinPay: {
                address: contracts.filecoin_pay_v1_addr,
                abi: Abis.filecoinPay,
            },
            fwss: {
                address: contracts.fwss_service_proxy_addr,
                abi: Abis.fwss,
            },
            fwssView: {
                address: contracts.fwss_state_view_addr,
                abi: Abis.fwssView,
            },
            serviceProviderRegistry: {
                address: contracts.service_provider_registry_proxy_addr,
                abi: Abis.serviceProviderRegistry,
            },
            sessionKeyRegistry: {
                address: contracts.session_key_registry_addr,
                abi: Abis.sessionKeyRegistry,
            },
            pdp: {
                address: contracts.pdp_verifier_proxy_addr,
                abi: Abis.pdp,
            },
            endorsements: {
                address: contracts.endorsements_addr,
                abi: Abis.providerIdSet,
            },
        },
        filbeam: null,
        testnet: true,
        genesisTimestamp: 0,
    };
}
export { validateDevnetInfo } from "./schema.js";
