// @filoz/synapse-core/src/sp/sp.ts
import { type AbortError, HttpError, type NetworkError, request, TimeoutError } from 'iso-web/http'
import type { ToString } from 'multiformats'
import { type Address, type Hex, isHex } from 'viem'
import {
  AddPiecesError,
  CreateDataSetError,
  DeletePieceError,
  type DownloadPieceError,
  FindPieceError,
  InvalidUploadSizeError,
  LocationHeaderError,
  PostPieceError,
  UploadPieceError,
} from '../errors/pdp.ts'
import { PullError } from '../errors/pull.ts'
import type { PieceCID } from '../piece/piece.ts'
import * as Piece from '../piece/piece.ts'
import type * as TypedData from '../typed-data/index.ts'
import { RETRY_CONSTANTS, SIZE_CONSTANTS } from '../utils/constants.ts'
import { isUint8Array } from '../utils/streams.ts'

export namespace createDataSet {
  /**
   * The options for the create data set on PDP API.
   */
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The address of the record keeper. */
    recordKeeper: Address
    /** The extra data for the create data set. */
    extraData: Hex
  }

  export type OutputType = {
    txHash: Hex
    statusUrl: string
  }

  export type ErrorType = CreateDataSetError | LocationHeaderError | TimeoutError | NetworkError | AbortError

  export type RequestBody = {
    recordKeeper: Address
    extraData: Hex
  }
}

/**
 * Create a data set on PDP API
 *
 * POST /pdp/data-sets
 *
 * @param options - {@link createDataSet.OptionsType}
 * @returns Transaction hash and status URL. {@link createDataSet.OutputType}
 * @throws Errors {@link createDataSet.ErrorType}
 */
export async function createDataSet(options: createDataSet.OptionsType): Promise<createDataSet.OutputType> {
  // Send the create data set message to the PDP
  const response = await request.post(new URL(`pdp/data-sets`, options.serviceURL), {
    body: JSON.stringify({
      recordKeeper: options.recordKeeper,
      extraData: options.extraData,
    }),
    headers: {
      'Content-Type': 'application/json',
    },
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new CreateDataSetError(await response.error.response.text())
    }
    throw response.error
  }

  const location = response.result.headers.get('Location')
  const hash = location?.split('/').pop()
  if (!location || !hash || !isHex(hash)) {
    throw new LocationHeaderError(location)
  }

  return {
    txHash: hash,
    statusUrl: new URL(location, options.serviceURL).toString(),
  }
}

export namespace createDataSetAndAddPieces {
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The address of the record keeper. */
    recordKeeper: Address
    /** The extra data for the create data set and add pieces. */
    extraData: Hex
    /** The pieces to add. */
    pieces: PieceCID[]
  }
  export type OutputType = {
    /** The transaction hash. */
    txHash: Hex
    /** The status URL. */
    statusUrl: string
  }
  export type ErrorType = CreateDataSetError | LocationHeaderError | TimeoutError | NetworkError | AbortError
  export type RequestBody = {
    recordKeeper: Address
    extraData: Hex
    pieces: {
      pieceCid: ToString<PieceCID>
      subPieces: { subPieceCid: ToString<PieceCID> }[]
    }[]
  }
}

/**
 * Create a data set and add pieces to it on PDP API
 *
 * POST /pdp/data-sets/create-and-add
 *
 * @param options - {@link createDataSetAndAddPieces.OptionsType}
 * @returns Hash and status URL {@link createDataSetAndAddPieces.OutputType}
 * @throws Errors {@link createDataSetAndAddPieces.ErrorType}
 */
export async function createDataSetAndAddPieces(
  options: createDataSetAndAddPieces.OptionsType
): Promise<createDataSetAndAddPieces.OutputType> {
  // Send the create data set message to the PDP
  const response = await request.post(new URL(`pdp/data-sets/create-and-add`, options.serviceURL), {
    body: JSON.stringify({
      recordKeeper: options.recordKeeper,
      extraData: options.extraData,
      pieces: options.pieces.map((piece) => ({
        pieceCid: piece.toString(),
        subPieces: [{ subPieceCid: piece.toString() }],
      })),
    }),
    headers: {
      'Content-Type': 'application/json',
    },
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new CreateDataSetError(await response.error.response.text())
    }
    throw response.error
  }

  const location = response.result.headers.get('Location')
  const hash = location?.split('/').pop()
  if (!location || !hash || !isHex(hash)) {
    throw new LocationHeaderError(location)
  }

  return {
    txHash: hash,
    statusUrl: new URL(location, options.serviceURL).toString(),
  }
}

export namespace uploadPiece {
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The data to upload. */
    data: Uint8Array
    /** The piece CID to upload. */
    pieceCid: PieceCID
  }
  export type ErrorType = InvalidUploadSizeError | LocationHeaderError | TimeoutError | NetworkError | AbortError
}

/**
 * Upload a piece to the PDP API.
 *
 * POST /pdp/piece
 *
 * @param options - {@link uploadPiece.OptionsType}
 * @throws Errors {@link uploadPiece.ErrorType}
 */
export async function uploadPiece(options: uploadPiece.OptionsType): Promise<void> {
  const size = options.data.length
  if (size < SIZE_CONSTANTS.MIN_UPLOAD_SIZE || size > SIZE_CONSTANTS.MAX_UPLOAD_SIZE) {
    throw new InvalidUploadSizeError(size)
  }

  const pieceCid = options.pieceCid
  if (!Piece.isPieceCID(pieceCid)) {
    throw new Error(`Invalid PieceCID: ${String(options.pieceCid)}`)
  }
  const response = await request.post(new URL(`pdp/piece`, options.serviceURL), {
    body: JSON.stringify({
      pieceCid: pieceCid.toString(),
    }),
    headers: {
      'Content-Type': 'application/json',
    },
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new PostPieceError(await response.error.response.text())
    }
    throw response.error
  }
  if (response.result.status === 200) {
    // Piece already exists on server
    return
  }

  // Extract upload ID from Location header
  const location = response.result.headers.get('Location')
  const uploadUuid = location?.split('/').pop()
  if (!location || !uploadUuid) {
    throw new LocationHeaderError(location)
  }

  const uploadResponse = await request.put(new URL(`pdp/piece/upload/${uploadUuid}`, options.serviceURL), {
    body: options.data as BufferSource,
    headers: {
      'Content-Type': 'application/octet-stream',
      'Content-Length': options.data.length.toString(),
    },
    timeout: false,
  })

  if (uploadResponse.error) {
    if (HttpError.is(uploadResponse.error)) {
      throw new UploadPieceError(await uploadResponse.error.response.text())
    }
    throw uploadResponse.error
  }
}

export type UploadPieceStreamingData = Uint8Array | ReadableStream | import('node:stream/web').ReadableStream
export namespace uploadPieceStreaming {
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The data to upload. */
    data: UploadPieceStreamingData
    /** The size of the data. If defined, it will be used to set the Content-Length header. */
    size?: number
    /** The progress callback. */
    onProgress?: (bytesUploaded: number) => void
    /** The piece CID to upload. */
    pieceCid?: PieceCID
    /** The signal to abort the request. */
    signal?: AbortSignal
  }
  export type OutputType = {
    pieceCid: PieceCID
    size: number
  }

  export type ErrorType = InvalidUploadSizeError | PostPieceError | LocationHeaderError
}

/**
 * Upload piece data using the 3-step CommP-last streaming protocol.
 *
 * Protocol:
 * 1. POST /pdp/piece/uploads → get upload session UUID
 * 2. PUT /pdp/piece/uploads/{uuid} → stream data while calculating CommP
 * 3. POST /pdp/piece/uploads/{uuid} → finalize with calculated CommP
 *
 * @param options - {@link uploadPieceStreaming.OptionsType}
 * @returns PieceCID and size of uploaded data {@link uploadPieceStreaming.OutputType}
 * @throws Errors {@link uploadPieceStreaming.ErrorType}
 */
export async function uploadPieceStreaming(
  options: uploadPieceStreaming.OptionsType
): Promise<uploadPieceStreaming.OutputType> {
  // Create upload session (POST /pdp/piece/uploads)
  const createResponse = await request.post(new URL('pdp/piece/uploads', options.serviceURL), {
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    signal: options.signal,
  })

  if (createResponse.error) {
    if (HttpError.is(createResponse.error)) {
      throw new PostPieceError(`Failed to create upload session: ${await createResponse.error.response.text()}`)
    }
    throw createResponse.error
  }

  if (createResponse.result.status !== 201) {
    throw new PostPieceError(`Expected 201 Created, got ${createResponse.result.status}`)
  }

  // Extract UUID from Location header: /pdp/piece/uploads/{uuid}
  const location = createResponse.result.headers.get('Location')
  if (!location) {
    throw new LocationHeaderError('Upload session created but Location header missing')
  }

  const locationMatch = location.match(/\/pdp\/piece\/uploads\/([a-fA-F0-9-]+)/)
  if (!locationMatch) {
    throw new LocationHeaderError(`Invalid Location header format: ${location}`)
  }

  const uploadUuid = locationMatch[1]

  // Create CommP calculator stream only if PieceCID not provided
  let getPieceCID: () => PieceCID | null = () => options.pieceCid ?? null
  let pieceCidStream: TransformStream<Uint8Array, Uint8Array> | null = null

  if (options.pieceCid == null) {
    const result = Piece.createPieceCIDStream()
    pieceCidStream = result.stream
    getPieceCID = result.getPieceCID
  }

  const dataStream = isUint8Array(options.data)
    ? new Blob([options.data as Uint8Array<ArrayBuffer>]).stream()
    : (options.data as ReadableStream) // ReadableStream types dont match between browsers and Node.js

  const size = isUint8Array(options.data) ? options.data.length : options.size

  // Add size tracking and progress reporting
  let bytesUploaded = 0
  const trackingStream = new TransformStream<unknown, Uint8Array>({
    transform(chunk, controller) {
      let bytes: Uint8Array | undefined

      if (isUint8Array(chunk)) {
        bytes = chunk
      } else if (ArrayBuffer.isView(chunk)) {
        bytes = new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength)
      } else {
        controller.error('Invalid chunk type only Uint8Array and TypedArray are supported')
        return
      }

      bytesUploaded += bytes.length

      if (bytesUploaded > SIZE_CONSTANTS.MAX_UPLOAD_SIZE) {
        controller.error(new InvalidUploadSizeError(bytesUploaded))
        return
      }

      // Report progress if callback provided
      if (options.onProgress) {
        options.onProgress(bytesUploaded)
      }

      controller.enqueue(bytes)
    },
    flush(controller) {
      if (bytesUploaded < SIZE_CONSTANTS.MIN_UPLOAD_SIZE) {
        controller.error(new InvalidUploadSizeError(bytesUploaded))
        return
      }
    },
  })

  // Chain streams: data → tracking → CommP calculation (if needed)
  const bodyStream = pieceCidStream
    ? dataStream.pipeThrough(trackingStream).pipeThrough(pieceCidStream)
    : dataStream.pipeThrough(trackingStream)

  // PUT /pdp/piece/uploads/{uuid} with streaming body
  const headers: Record<string, string> = {
    'Content-Type': 'application/octet-stream',
    ...(size != null ? { 'Content-Length': size.toString() } : {}),
  }

  const uploadResponse = await request.put(new URL(`pdp/piece/uploads/${uploadUuid}`, options.serviceURL), {
    body: bodyStream,
    headers,
    timeout: false, // No timeout for streaming upload
    signal: options.signal,
    duplex: 'half', // Required for streaming request bodies
  } as Parameters<typeof request.put>[1] & { duplex: 'half' })

  if (uploadResponse.error) {
    if (HttpError.is(uploadResponse.error)) {
      throw new UploadPieceError(`Failed to upload piece: ${await uploadResponse.error.response.text()}`)
    }
    throw uploadResponse.error
  }

  if (uploadResponse.result.status !== 204) {
    throw new UploadPieceError(`Expected 204 No Content, got ${uploadResponse.result.status}`)
  }

  // Get PieceCID (either provided or calculated) and finalize
  const pieceCid = getPieceCID()
  if (pieceCid === null) {
    throw new Error('Failed to calculate PieceCID during upload')
  }

  const finalizeBody = JSON.stringify({
    pieceCid: pieceCid.toString(),
  })

  // POST /pdp/piece/uploads/{uuid} with PieceCID
  const finalizeResponse = await request.post(new URL(`pdp/piece/uploads/${uploadUuid}`, options.serviceURL), {
    body: finalizeBody,
    headers: {
      'Content-Type': 'application/json',
    },
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    signal: options.signal,
  })

  if (finalizeResponse.error) {
    if (HttpError.is(finalizeResponse.error)) {
      throw new PostPieceError(`Failed to finalize upload: ${await finalizeResponse.error.response.text()}`)
    }
    throw finalizeResponse.error
  }

  if (finalizeResponse.result.status !== 200) {
    throw new PostPieceError(`Expected 200 OK for finalization, got ${finalizeResponse.result.status}`)
  }

  return {
    pieceCid,
    size: bytesUploaded,
  }
}

export namespace findPiece {
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The piece CID to find. */
    pieceCid: PieceCID
    /** The signal to abort the request. */
    signal?: AbortSignal
    /** Whether to retry the request. Defaults to false. */
    retry?: boolean
    /** The timeout in milliseconds. Defaults to 5 minutes. */
    timeout?: number
    /** The poll interval in milliseconds. Defaults to 1 second. */
    pollInterval?: number
  }
  export type OutputType = PieceCID
  export type ErrorType = FindPieceError | TimeoutError | NetworkError | AbortError
}
/**
 * Find a piece on the PDP API.
 *
 * GET /pdp/piece?pieceCid={pieceCid}
 *
 * @param options - {@link findPiece.OptionsType}
 * @returns Piece CID {@link findPiece.OutputType}
 * @throws Errors {@link findPiece.ErrorType}
 */
export async function findPiece(options: findPiece.OptionsType): Promise<findPiece.OutputType> {
  const { pieceCid, serviceURL } = options
  const params = new URLSearchParams({ pieceCid: pieceCid.toString() })
  const retry = options.retry ?? false
  const response = await request.json.get<{ pieceCid: string }>(new URL(`pdp/piece?${params.toString()}`, serviceURL), {
    signal: options.signal,
    retry: retry
      ? {
          statusCodes: [202, 404],
          retries: RETRY_CONSTANTS.RETRIES,
          factor: RETRY_CONSTANTS.FACTOR,
          minTimeout: options.pollInterval ?? 1000,
        }
      : undefined,
    timeout: options.timeout ?? RETRY_CONSTANTS.MAX_RETRY_TIME,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new FindPieceError(await response.error.response.text())
    }
    if (TimeoutError.is(response.error)) {
      throw new FindPieceError('Timeout waiting for piece to be found')
    }
    throw response.error
  }
  const data = response.result
  return Piece.parse(data.pieceCid)
}

export namespace addPieces {
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The ID of the data set. */
    dataSetId: bigint
    /** The pieces to add. */
    pieces: PieceCID[]
    /** The extra data for the add pieces. {@link TypedData.signAddPieces} */
    extraData: Hex
  }
  export type OutputType = {
    /** The transaction hash. */
    txHash: Hex
    /** The status URL. */
    statusUrl: string
  }
  export type ErrorType = AddPiecesError | LocationHeaderError | TimeoutError | NetworkError | AbortError
  export type RequestBody = {
    pieces: {
      pieceCid: ToString<PieceCID>
      subPieces: { subPieceCid: ToString<PieceCID> }[]
    }[]
    extraData: Hex
  }
}

/**
 * Add pieces to a data set on the PDP API.
 *
 * POST /pdp/data-sets/{dataSetId}/pieces
 *
 * @param options - {@link addPieces.OptionsType}
 * @returns Hash and status URL {@link addPieces.OutputType}
 * @throws Errors {@link addPieces.ErrorType}
 */
export async function addPieces(options: addPieces.OptionsType): Promise<addPieces.OutputType> {
  const { serviceURL, dataSetId, pieces, extraData } = options
  const response = await request.post(new URL(`pdp/data-sets/${dataSetId}/pieces`, serviceURL), {
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      pieces: pieces.map((piece) => ({
        pieceCid: piece.toString(),
        subPieces: [{ subPieceCid: piece.toString() }],
      })),
      extraData: extraData,
    }),
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new AddPiecesError(await response.error.response.text())
    }
    throw response.error
  }
  const location = response.result.headers.get('Location')
  const txHash = location?.split('/').pop()
  if (!location || !txHash || !isHex(txHash)) {
    throw new LocationHeaderError(location)
  }

  return {
    txHash: txHash as Hex,
    statusUrl: new URL(location, serviceURL).toString(),
  }
}

export namespace deletePiece {
  export type OptionsType = {
    serviceURL: string
    dataSetId: bigint
    pieceId: bigint
    extraData: Hex
  }
  export type OutputType = {
    hash: Hex
  }
  export type ErrorType = DeletePieceError | TimeoutError | NetworkError | AbortError
}

/**
 * Delete a piece from a data set on the PDP API.
 *
 * DELETE /pdp/data-sets/{dataSetId}/pieces/{pieceId}
 *
 * @param options - {@link deletePiece.OptionsType}
 * @returns Hash of the delete operation {@link deletePiece.OutputType}
 * @throws Errors {@link deletePiece.ErrorType}
 */
export async function deletePiece(options: deletePiece.OptionsType): Promise<deletePiece.OutputType> {
  const { serviceURL, dataSetId, pieceId, extraData } = options
  const response = await request.json.delete<{ txHash: Hex }>(
    new URL(`pdp/data-sets/${dataSetId}/pieces/${pieceId}`, serviceURL),
    {
      body: { extraData },
      timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    }
  )

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new DeletePieceError(await response.error.response.text())
    }
    throw response.error
  }

  return { hash: response.result.txHash }
}

/**
 * Ping the PDP API.
 *
 * GET /pdp/ping
 *
 * @param serviceURL - The service URL of the PDP API.
 * @returns void
 * @throws Errors {@link Error}
 */
export async function ping(serviceURL: string) {
  const response = await request.get(new URL(`pdp/ping`, serviceURL))
  if (response.error) {
    throw new Error('Ping failed')
  }
  return response.result
}

export namespace downloadPiece {
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The piece CID to download. */
    pieceCid: PieceCID
  }
  export type ReturnType = Uint8Array
  export type ErrorType = DownloadPieceError | TimeoutError | NetworkError | AbortError
}

// =============================================================================
// SP-to-SP Piece Pull Operations
// =============================================================================

/**
 * Status of a pull operation or individual piece.
 *
 * Status progression:
 * - `pending`: Piece is queued but download hasn't started
 * - `inProgress`: Download task is actively running (first attempt)
 * - `retrying`: Download task is running after one or more failures
 * - `complete`: Piece successfully downloaded and verified
 * - `failed`: Piece permanently failed after exhausting retries
 *
 * Overall response status reflects the worst-case across all pieces:
 * failed > retrying > inProgress > pending > complete
 */
export type PullStatus = 'pending' | 'inProgress' | 'retrying' | 'complete' | 'failed'

/**
 * Input piece for a pull request.
 */
export type PullPieceInput = {
  /** PieceCIDv2 format (encodes both CommP and raw size) */
  pieceCid: string
  /** HTTPS URL to pull the piece from (must end in /piece/{pieceCid}) */
  sourceUrl: string
}

/**
 * Status of a single piece in a pull response.
 */
export type PullPieceStatus = {
  /** PieceCIDv2 of the piece */
  pieceCid: string
  /** Current status of this piece */
  status: PullStatus
}

/**
 * Response from a pull request.
 */
export type PullResponse = {
  /** Overall status (worst-case across all pieces) */
  status: PullStatus
  /** Per-piece status */
  pieces: PullPieceStatus[]
}

export namespace pullPieces {
  /**
   * Options for pulling pieces from external SPs.
   */
  export type OptionsType = {
    /** The service URL of the PDP API. */
    serviceURL: string
    /** The record keeper contract address (e.g., FWSS). */
    recordKeeper: Address
    /** EIP-712 signed extraData for authorization. */
    extraData: Hex
    /** Optional target dataset ID (omit or 0n to create new). */
    dataSetId?: bigint
    /** Pieces to pull with their source URLs. */
    pieces: PullPieceInput[]
    /** Optional AbortSignal to cancel the request. */
    signal?: AbortSignal
  }

  export type ReturnType = PullResponse

  export type ErrorType = PullError | TimeoutError | NetworkError | AbortError

  export type RequestBody = {
    extraData: Hex
    recordKeeper: Address
    pieces: PullPieceInput[]
    dataSetId?: number
  }
}

/**
 * Build the JSON request body for a pull request.
 */
function buildPullRequestBody(options: pullPieces.OptionsType): string {
  const body: pullPieces.RequestBody = {
    extraData: options.extraData,
    recordKeeper: options.recordKeeper,
    pieces: options.pieces,
  }

  // Only include dataSetId if specified and non-zero
  if (options.dataSetId != null && options.dataSetId > 0n) {
    body.dataSetId = Number(options.dataSetId)
  }

  return JSON.stringify(body)
}

/**
 * Initiate a piece pull request or get status of an existing one.
 *
 * POST /pdp/piece/pull
 *
 * This endpoint is idempotent - calling with the same extraData returns
 * the status of the existing pull rather than creating duplicates.
 * This allows safe retries and status polling using the same request.
 *
 * @param options - {@link pullPieces.OptionsType}
 * @returns The current status of the pull operation. {@link pullPieces.ReturnType}
 * @throws Errors {@link pullPieces.ErrorType}
 */
export async function pullPieces(options: pullPieces.OptionsType): Promise<pullPieces.ReturnType> {
  const response = await request.post(new URL('pdp/piece/pull', options.serviceURL), {
    body: buildPullRequestBody(options),
    headers: {
      'Content-Type': 'application/json',
    },
    timeout: RETRY_CONSTANTS.MAX_RETRY_TIME,
    signal: options.signal,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new PullError(await response.error.response.text())
    }
    throw response.error
  }

  return (await response.result.json()) as pullPieces.ReturnType
}

export namespace waitForPullStatus {
  /**
   * Options for polling pull status.
   */
  export type OptionsType = pullPieces.OptionsType & {
    /** Callback invoked on each poll with current status. */
    onStatus?: (response: PullResponse) => void
    /** The timeout in milliseconds. Defaults to 5 minutes. */
    timeout?: number
    /** The polling interval in milliseconds. Defaults to 4 seconds. */
    pollInterval?: number
  }

  export type ReturnType = PullResponse

  export type ErrorType = PullError | TimeoutError | NetworkError | AbortError
}

/**
 * Wait for pull completion.
 *
 * Repeatedly calls the pull endpoint until all pieces are complete or any piece fails.
 * Since the endpoint is idempotent, this effectively polls for status updates.
 *
 * @param options - {@link waitForPullStatus.OptionsType}
 * @returns The final status when complete or failed. {@link waitForPullStatus.ReturnType}
 * @throws Errors {@link waitForPullStatus.ErrorType}
 */
export async function waitForPullStatus(options: waitForPullStatus.OptionsType): Promise<waitForPullStatus.ReturnType> {
  const url = new URL('pdp/piece/pull', options.serviceURL)
  const body = buildPullRequestBody(options)
  const headers = { 'Content-Type': 'application/json' }

  // Custom fetch that creates a fresh Request each time to avoid body consumption issues
  // (iso-web creates Request once and reuses it, but POST bodies can only be read once)
  const fetchWithFreshRequest: typeof globalThis.fetch = (input, init) => {
    // iso-web passes the Request object as input, extract signal from it
    const signal = input instanceof Request ? input.signal : init?.signal
    return globalThis.fetch(url, { method: 'POST', body, headers, signal })
  }

  const response = await request.post(url, {
    body,
    headers,
    fetch: fetchWithFreshRequest,
    async onResponse(response) {
      if (response.ok) {
        const data = (await response.clone().json()) as PullResponse

        // Invoke status callback if provided
        if (options.onStatus) {
          options.onStatus(data)
        }

        // Stop polling when complete or failed
        if (data.status === 'complete' || data.status === 'failed') {
          return response
        }
        throw new Error('Pull not complete')
      }
    },
    retry: {
      shouldRetry: (ctx) => ctx.error.message === 'Pull not complete',
      retries: RETRY_CONSTANTS.RETRIES,
      factor: RETRY_CONSTANTS.FACTOR,
      minTimeout: options.pollInterval ?? RETRY_CONSTANTS.DELAY_TIME,
    },
    timeout: options.timeout ?? RETRY_CONSTANTS.MAX_RETRY_TIME,
    signal: options.signal,
  })

  if (response.error) {
    if (HttpError.is(response.error)) {
      throw new PullError(await response.error.response.text())
    }
    throw response.error
  }

  return (await response.result.json()) as waitForPullStatus.ReturnType
}
