// @filoz/synapse-core/dist/src/sp-registry/get-pdp-provider.js
import { decodePDPOffering } from "../utils/pdp-capabilities.js";
import { getProviderIdByAddress } from "./get-provider-id-by-address.js";
import { getProviderWithProduct, getProviderWithProductCall } from "./get-provider-with-product.js";
import { PRODUCTS } from "./types.js";
export async function getPDPProvider(client, options) {
    const data = await getProviderWithProduct(client, {
        ...options,
        productType: PRODUCTS.PDP,
    });
    return parsePDPProvider(data);
}
export function getPDPProviderCall(options) {
    return getProviderWithProductCall({
        ...options,
        productType: PRODUCTS.PDP,
    });
}
export function parsePDPProvider(data) {
    return {
        id: data.providerId,
        ...data.providerInfo,
        pdp: decodePDPOffering(data),
    };
}
export async function getPDPProviderByAddress(client, options) {
    const providerId = await getProviderIdByAddress(client, {
        providerAddress: options.address,
        contractAddress: options.contractAddress,
    });
    if (providerId === 0n) {
        return null;
    }
    return getPDPProvider(client, { providerId, contractAddress: options.contractAddress });
}
