// @filoz/synapse-sdk/dist/src/storage/context.js
import { asChain } from '@filoz/synapse-core/chains';
import { InvalidPieceCIDError } from '@filoz/synapse-core/errors';
import * as PDPVerifier from '@filoz/synapse-core/pdp-verifier';
import * as Piece from '@filoz/synapse-core/piece';
import * as SP from '@filoz/synapse-core/sp';
import { schedulePieceDeletion } from '@filoz/synapse-core/sp';
import { signAddPieces, signCreateDataSetAndAddPieces } from '@filoz/synapse-core/typed-data';
import { calculateLastProofDate, createPieceUrlPDP, datasetMetadataObjectToEntry, epochToDate, pieceMetadataObjectToEntry, randU256, timeUntilEpoch, } from '@filoz/synapse-core/utils';
import { fetchProviderSelectionInput, metadataMatches, selectProviders, } from '@filoz/synapse-core/warm-storage';
import { getBlockNumber } from 'viem/actions';
import { SPRegistryService } from "../sp-registry/index.js";
import { createError, SIZE_CONSTANTS } from "../utils/index.js";
import { combineMetadata } from "../utils/metadata.js";
const NO_REMAINING_PROVIDERS_ERROR_MESSAGE = 'No approved service providers available';
export class StorageContext {
    _client;
    _chain;
    _synapse;
    _provider;
    _pdpEndpoint;
    _warmStorageService;
    _withCDN;
    _dataSetId;
    _clientDataSetId;
    _dataSetMetadata;
    serviceProvider;
    get withCDN() {
        return this._withCDN;
    }
    get provider() {
        return this._provider;
    }
    get dataSetMetadata() {
        return this._dataSetMetadata;
    }
    get dataSetId() {
        return this._dataSetId;
    }
    async getClientDataSetId() {
        if (this._clientDataSetId !== undefined) {
            return this._clientDataSetId;
        }
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'getClientDataSetId', 'Data set not found');
        }
        const dataSetInfo = await this._warmStorageService.getDataSet({ dataSetId: this.dataSetId });
        if (dataSetInfo == null) {
            throw createError('StorageContext', 'getClientDataSetId', 'Data set not found');
        }
        this._clientDataSetId = dataSetInfo.clientDataSetId;
        return this._clientDataSetId;
    }
    static validateRawSize(options) {
        const { sizeBytes, context } = options;
        if (sizeBytes < SIZE_CONSTANTS.MIN_UPLOAD_SIZE) {
            throw createError('StorageContext', context, `Data size ${sizeBytes} bytes is below minimum allowed size of ${SIZE_CONSTANTS.MIN_UPLOAD_SIZE} bytes`);
        }
        if (sizeBytes > SIZE_CONSTANTS.MAX_UPLOAD_SIZE) {
            throw createError('StorageContext', context, `Data size ${sizeBytes} bytes exceeds maximum allowed size of ${SIZE_CONSTANTS.MAX_UPLOAD_SIZE} bytes (${Math.floor(SIZE_CONSTANTS.MAX_UPLOAD_SIZE / 1024 / 1024)} MiB)`);
        }
    }
    constructor(options) {
        this._client = options.synapse.client;
        this._chain = asChain(this._client.chain);
        this._synapse = options.synapse;
        this._provider = options.provider;
        this._withCDN = options.options.withCDN ?? false;
        this._warmStorageService = options.warmStorageService;
        this._dataSetMetadata = options.dataSetMetadata;
        this._dataSetId = options.dataSetId;
        this.serviceProvider = options.provider.serviceProvider;
        this._pdpEndpoint = options.provider.pdp.serviceURL;
    }
    static async createContexts(options) {
        const clientAddress = options.synapse.client.account.address;
        const spRegistry = new SPRegistryService({ client: options.synapse.client });
        const hasDataSetIds = options.dataSetIds != null && options.dataSetIds.length > 0;
        const hasProviderIds = options.providerIds != null && options.providerIds.length > 0;
        if (hasDataSetIds && hasProviderIds) {
            throw createError('StorageContext', 'createContexts', "Cannot specify both 'dataSetIds' and 'providerIds' - use one or the other");
        }
        let resolutions = [];
        if (hasDataSetIds) {
            const uniqueDataSetIds = [...new Set(options.dataSetIds)];
            resolutions = await Promise.all(uniqueDataSetIds.map((dataSetId) => StorageContext.resolveByDataSetId(dataSetId, options.warmStorageService, spRegistry, clientAddress)));
        }
        else if (hasProviderIds) {
            const uniqueProviderIds = [...new Set(options.providerIds)];
            resolutions = await Promise.all(uniqueProviderIds.map((providerId) => StorageContext.resolveByProviderId(clientAddress, providerId, options.metadata ?? {}, options.warmStorageService, spRegistry)));
        }
        if (resolutions.length > 0) {
            const count = options.count ?? resolutions.length;
            if (resolutions.length !== count) {
                throw createError('StorageContext', 'createContexts', `Requested ${count} context(s) but ${hasDataSetIds ? 'dataSetIds' : 'providerIds'}` +
                    ` resolved to ${resolutions.length} after deduplication`);
            }
            if (hasDataSetIds && resolutions.length > 1) {
                const providerIds = resolutions.map((r) => r.provider.id);
                if (new Set(providerIds).size !== providerIds.length) {
                    throw createError('StorageContext', 'createContexts', 'dataSetIds resolve to duplicate providers - each context must use a unique provider');
                }
            }
        }
        else {
            const count = options.count ?? 2;
            resolutions = await StorageContext.smartSelect({
                synapse: options.synapse,
                metadata: options.metadata ?? {},
                count,
                excludeProviderIds: new Set(options.excludeProviderIds ?? []),
                requireEndorsedPrimary: true,
            });
        }
        return await Promise.all(resolutions.map(async (resolution) => await StorageContext.createWithSelectedProvider(resolution, options.synapse, options.warmStorageService, options)));
    }
    static async create(options) {
        const spRegistry = new SPRegistryService({ client: options.synapse.client });
        const resolution = await StorageContext.resolveProviderAndDataSet(options.synapse, options.warmStorageService, spRegistry, options);
        return await StorageContext.createWithSelectedProvider(resolution, options.synapse, options.warmStorageService, options);
    }
    static async createWithSelectedProvider(resolution, synapse, warmStorageService, options = {}) {
        try {
            options.callbacks?.onProviderSelected?.(resolution.provider);
        }
        catch (error) {
            console.error('Error in onProviderSelected callback:', error);
        }
        if (resolution.dataSetId != null) {
            options.callbacks?.onDataSetResolved?.({
                dataSetId: resolution.dataSetId,
                provider: resolution.provider,
            });
        }
        return new StorageContext({
            synapse,
            warmStorageService,
            provider: resolution.provider,
            dataSetId: resolution.dataSetId ?? undefined,
            options,
            dataSetMetadata: resolution.dataSetMetadata,
        });
    }
    static async resolveProviderAndDataSet(synapse, warmStorageService, spRegistry, options) {
        const clientAddress = synapse.client.account.address;
        const requestedMetadata = combineMetadata(options.metadata, options.withCDN);
        if (options.dataSetId != null) {
            const result = await StorageContext.resolveByDataSetId(options.dataSetId, warmStorageService, spRegistry, clientAddress);
            if (options.providerId != null && result.provider.id !== options.providerId) {
                throw createError('StorageContext', 'resolveProviderAndDataSet', `Data set ${options.dataSetId} belongs to provider ID ${result.provider.id}, but provider ID ${options.providerId} was requested`);
            }
            return result;
        }
        if (options.providerId != null) {
            return await StorageContext.resolveByProviderId(clientAddress, options.providerId, requestedMetadata, warmStorageService, spRegistry);
        }
        const results = await StorageContext.smartSelect({
            synapse,
            metadata: requestedMetadata,
            count: 1,
            excludeProviderIds: new Set(options.excludeProviderIds ?? []),
            requireEndorsedPrimary: false,
        });
        if (results.length === 0) {
            throw createError('StorageContext', 'resolveProviderAndDataSet', NO_REMAINING_PROVIDERS_ERROR_MESSAGE);
        }
        return results[0];
    }
    static async resolveByDataSetId(dataSetId, warmStorageService, spRegistry, clientAddress) {
        const [dataSetInfo, dataSetMetadata] = await Promise.all([
            warmStorageService.getDataSet({ dataSetId }),
            warmStorageService.getDataSetMetadata({ dataSetId }),
            warmStorageService.validateDataSet({ dataSetId }),
        ]);
        if (dataSetInfo == null) {
            throw createError('StorageContext', 'resolveByDataSetId', `Data set ${dataSetId} does not exist`);
        }
        if (dataSetInfo.payer.toLowerCase() !== clientAddress.toLowerCase()) {
            throw createError('StorageContext', 'resolveByDataSetId', `Data set ${dataSetId} is not owned by ${clientAddress} (owned by ${dataSetInfo.payer})`);
        }
        const provider = await spRegistry.getProvider({ providerId: dataSetInfo.providerId });
        if (provider == null) {
            throw createError('StorageContext', 'resolveByDataSetId', `Provider ID ${dataSetInfo.providerId} for data set ${dataSetId} not found in registry`);
        }
        return {
            provider,
            dataSetId,
            dataSetMetadata,
        };
    }
    static async resolveByProviderId(clientAddress, providerId, requestedMetadata, warmStorageService, spRegistry) {
        const [provider, dataSets] = await Promise.all([
            spRegistry.getProvider({ providerId }),
            warmStorageService.getClientDataSets({ address: clientAddress }),
        ]);
        if (provider == null) {
            throw createError('StorageContext', 'resolveByProviderId', `Provider ID ${providerId} not found in registry`);
        }
        const providerDataSets = dataSets.filter((dataSet) => dataSet.dataSetId && dataSet.providerId === provider.id && dataSet.pdpEndEpoch === 0n);
        const sortedDataSets = providerDataSets.sort((a, b) => {
            return Number(a.dataSetId) - Number(b.dataSetId);
        });
        const MIN_BATCH_SIZE = 50;
        const MAX_BATCH_SIZE = 200;
        const BATCH_SIZE = Math.min(MAX_BATCH_SIZE, Math.max(MIN_BATCH_SIZE, Math.ceil(sortedDataSets.length / 3), 1));
        let selectedDataSet = null;
        for (let i = 0; i < sortedDataSets.length; i += BATCH_SIZE) {
            const batchResults = await Promise.all(sortedDataSets.slice(i, i + BATCH_SIZE).map(async (dataSet) => {
                const dataSetId = dataSet.dataSetId;
                try {
                    const [dataSetMetadata, activePieceCount] = await Promise.all([
                        warmStorageService.getDataSetMetadata({ dataSetId }),
                        warmStorageService.getActivePieceCount({ dataSetId }),
                        warmStorageService.validateDataSet({ dataSetId }),
                    ]);
                    if (!metadataMatches(dataSetMetadata, requestedMetadata)) {
                        return null;
                    }
                    return {
                        dataSetId,
                        dataSetMetadata,
                        activePieceCount,
                    };
                }
                catch (error) {
                    console.warn(`Skipping data set ${dataSetId} for provider ${providerId}:`, error instanceof Error ? error.message : String(error));
                    return null;
                }
            }));
            for (const result of batchResults) {
                if (result == null)
                    continue;
                if (result.activePieceCount > 0) {
                    selectedDataSet = result;
                    break;
                }
                if (selectedDataSet == null) {
                    selectedDataSet = result;
                }
            }
            if (selectedDataSet != null && selectedDataSet.activePieceCount > 0) {
                break;
            }
        }
        if (selectedDataSet != null) {
            return {
                provider,
                dataSetId: selectedDataSet.dataSetId,
                dataSetMetadata: selectedDataSet.dataSetMetadata,
            };
        }
        return {
            provider,
            dataSetId: null,
            dataSetMetadata: requestedMetadata,
        };
    }
    static async smartSelect(options) {
        const { synapse, metadata, count, requireEndorsedPrimary } = options;
        const clientAddress = synapse.client.account.address;
        const input = await fetchProviderSelectionInput(synapse.client, {
            address: clientAddress,
        });
        const results = [];
        const excludeProviderIds = new Set(options.excludeProviderIds);
        for (let i = 0; i < count; i++) {
            const endorsedSlot = requireEndorsedPrimary && i === 0;
            let found = false;
            let pingFailures = 0;
            for (;;) {
                const candidates = selectProviders({
                    ...input,
                    endorsedIds: endorsedSlot ? input.endorsedIds : new Set(),
                    count: 1,
                    excludeProviderIds,
                    metadata,
                });
                if (candidates.length === 0)
                    break;
                const candidate = candidates[0];
                try {
                    await SP.ping(candidate.provider.pdp.serviceURL);
                    results.push(StorageContext.toProviderSelectionResult(candidate));
                    excludeProviderIds.add(candidate.provider.id);
                    found = true;
                    break;
                }
                catch (error) {
                    console.warn(`Provider ${candidate.provider.serviceProvider} (ID: ${candidate.provider.id}) failed ping:`, error instanceof Error ? error.message : String(error));
                    excludeProviderIds.add(candidate.provider.id);
                    pingFailures++;
                }
            }
            if (endorsedSlot && !found) {
                throw createError('StorageContext', 'smartSelect', pingFailures > 0
                    ? `No endorsed provider available — all endorsed provider(s) failed health check`
                    : 'No endorsed provider available');
            }
            if (!found) {
                break;
            }
        }
        return results;
    }
    static toProviderSelectionResult(location) {
        return {
            provider: location.provider,
            dataSetId: location.dataSetId,
            dataSetMetadata: location.dataSetMetadata,
        };
    }
    static async performPreflightCheck(options) {
        const { size, withCDN, warmStorageService } = options;
        StorageContext.validateRawSize({ sizeBytes: options.size, context: 'preflightUpload' });
        const allowanceCheck = await warmStorageService.checkAllowanceForStorage({ sizeInBytes: BigInt(size), withCDN });
        return {
            estimatedCost: {
                perEpoch: allowanceCheck.costs.perEpoch,
                perDay: allowanceCheck.costs.perDay,
                perMonth: allowanceCheck.costs.perMonth,
            },
            allowanceCheck: {
                sufficient: allowanceCheck.sufficient,
                message: allowanceCheck.message,
            },
            selectedProvider: null,
            selectedDataSetId: null,
        };
    }
    async preflightUpload(options) {
        return await StorageContext.performPreflightCheck({
            size: options.size,
            withCDN: this._withCDN,
            warmStorageService: this._warmStorageService,
        });
    }
    async store(data, options) {
        if (data instanceof Uint8Array) {
            StorageContext.validateRawSize({ sizeBytes: data.length, context: 'store' });
        }
        let uploadResult;
        try {
            uploadResult = await SP.uploadPieceStreaming({
                serviceURL: this._pdpEndpoint,
                data,
                pieceCid: options?.pieceCid,
                signal: options?.signal,
                onProgress: options?.onProgress,
            });
        }
        catch (error) {
            throw createError('StorageContext', 'store', 'Failed to store piece on service provider', error);
        }
        try {
            await SP.findPiece({
                serviceURL: this._pdpEndpoint,
                pieceCid: uploadResult.pieceCid,
                retry: true,
                signal: options?.signal,
            });
        }
        catch (error) {
            throw createError('StorageContext', 'store', 'Failed to confirm piece storage', error);
        }
        return {
            pieceCid: uploadResult.pieceCid,
            size: uploadResult.size,
        };
    }
    async presignForCommit(pieces) {
        const signingPieces = pieces.map((p) => ({
            pieceCid: p.pieceCid,
            metadata: pieceMetadataObjectToEntry(p.pieceMetadata),
        }));
        if (this._dataSetId) {
            return signAddPieces(this._synapse.sessionClient ?? this._synapse.client, {
                clientDataSetId: await this.getClientDataSetId(),
                pieces: signingPieces,
            });
        }
        return signCreateDataSetAndAddPieces(this._synapse.sessionClient ?? this._synapse.client, {
            clientDataSetId: randU256(),
            payee: this._provider.serviceProvider,
            payer: this._synapse.client.account.address,
            metadata: datasetMetadataObjectToEntry(this._dataSetMetadata, {
                cdn: this._withCDN,
            }),
            pieces: signingPieces,
        });
    }
    async pull(options) {
        const { pieces, from, signal, onProgress, extraData } = options;
        const getSourceUrl = (pieceCid) => {
            if (typeof from === 'string') {
                return createPieceUrlPDP({ cid: pieceCid.toString(), serviceURL: from });
            }
            return from(pieceCid);
        };
        const pullPiecesInput = pieces.map((pieceCid) => ({
            pieceCid,
            sourceUrl: getSourceUrl(pieceCid),
        }));
        const handleProgressResponse = onProgress
            ? (response) => {
                for (const piece of response.pieces) {
                    const pieceCid = pieces.find((p) => p.toString() === piece.pieceCid);
                    if (pieceCid) {
                        onProgress(pieceCid, piece.status);
                    }
                }
            }
            : undefined;
        try {
            const sharedOptions = {
                serviceURL: this._pdpEndpoint,
                pieces: pullPiecesInput,
                signal,
                onStatus: handleProgressResponse,
                extraData,
            };
            const pullOptions = this._dataSetId
                ? {
                    ...sharedOptions,
                    dataSetId: this._dataSetId,
                    clientDataSetId: await this.getClientDataSetId(),
                }
                : {
                    ...sharedOptions,
                    payee: this._provider.serviceProvider,
                    payer: this._synapse.client.account.address,
                    cdn: this._withCDN,
                    metadata: this._dataSetMetadata,
                };
            const response = await SP.waitForPullStatus(this._synapse.sessionClient ?? this._synapse.client, pullOptions);
            const pieceResults = response.pieces.map((piece) => {
                const pieceCid = pieces.find((p) => p.toString() === piece.pieceCid);
                return {
                    pieceCid: pieceCid,
                    status: piece.status === 'complete' ? 'complete' : 'failed',
                };
            });
            const allComplete = pieceResults.every((p) => p.status === 'complete');
            return {
                status: allComplete ? 'complete' : 'failed',
                pieces: pieceResults,
            };
        }
        catch (error) {
            throw createError('StorageContext', 'pull', 'Failed to pull pieces from source provider', error);
        }
    }
    async commit(options) {
        const { pieces, extraData } = options;
        for (const piece of pieces) {
            if (piece.pieceMetadata) {
                pieceMetadataObjectToEntry(piece.pieceMetadata);
            }
        }
        const pieceInputs = pieces.map((p) => ({ pieceCid: p.pieceCid, metadata: p.pieceMetadata }));
        try {
            if (this._dataSetId) {
                const [, clientDataSetId] = await Promise.all([
                    this._warmStorageService.validateDataSet({ dataSetId: this._dataSetId }),
                    this.getClientDataSetId(),
                ]);
                const addPiecesResult = await SP.addPieces(this._synapse.sessionClient ?? this._client, {
                    dataSetId: this._dataSetId,
                    clientDataSetId,
                    pieces: pieceInputs,
                    serviceURL: this._pdpEndpoint,
                    extraData,
                });
                options.onSubmitted?.(addPiecesResult.txHash);
                const confirmation = await SP.waitForAddPieces(addPiecesResult);
                const confirmedPieceIds = confirmation.confirmedPieceIds;
                return {
                    txHash: addPiecesResult.txHash,
                    pieceIds: confirmedPieceIds,
                    dataSetId: this._dataSetId,
                    isNewDataSet: false,
                };
            }
            const result = await SP.createDataSetAndAddPieces(this._synapse.sessionClient ?? this._client, {
                cdn: this._withCDN,
                payee: this._provider.serviceProvider,
                payer: this._synapse.client.account.address,
                recordKeeper: this._chain.contracts.fwss.address,
                pieces: pieceInputs,
                metadata: this._dataSetMetadata,
                serviceURL: this._pdpEndpoint,
                extraData,
            });
            options.onSubmitted?.(result.txHash);
            const confirmation = await SP.waitForCreateDataSetAddPieces(result);
            this._dataSetId = confirmation.dataSetId;
            return {
                txHash: result.txHash,
                pieceIds: confirmation.piecesIds,
                dataSetId: this._dataSetId,
                isNewDataSet: true,
            };
        }
        catch (error) {
            throw createError('StorageContext', 'commit', 'Failed to commit pieces on-chain', error);
        }
    }
    getPieceUrl(pieceCid) {
        return createPieceUrlPDP({ cid: pieceCid.toString(), serviceURL: this._pdpEndpoint });
    }
    async upload(data, options) {
        const storeResult = await this.store(data, {
            pieceCid: options?.pieceCid,
            signal: options?.signal,
            onProgress: options?.onProgress,
        });
        options?.onStored?.(this._provider.id, storeResult.pieceCid);
        const commitResult = await this.commit({
            pieces: [{ pieceCid: storeResult.pieceCid, pieceMetadata: options?.pieceMetadata }],
            onSubmitted: (txHash) => options?.onPiecesAdded?.(txHash, this._provider.id, [{ pieceCid: storeResult.pieceCid }]),
        });
        const pieceId = commitResult.pieceIds[0];
        options?.onPiecesConfirmed?.(commitResult.dataSetId, this._provider.id, [
            { pieceId, pieceCid: storeResult.pieceCid },
        ]);
        return {
            pieceCid: storeResult.pieceCid,
            size: storeResult.size,
            copies: [
                {
                    providerId: this._provider.id,
                    dataSetId: commitResult.dataSetId,
                    pieceId,
                    role: 'primary',
                    retrievalUrl: this.getPieceUrl(storeResult.pieceCid),
                    isNewDataSet: commitResult.isNewDataSet,
                },
            ],
            failures: [],
        };
    }
    async download(options) {
        const parsedPieceCID = Piece.asPieceCID(options.pieceCid);
        if (parsedPieceCID == null) {
            throw new InvalidPieceCIDError(options.pieceCid);
        }
        const withCDN = options.withCDN ?? this._withCDN;
        const pieceUrl = await Piece.resolvePieceUrl({
            client: this._client,
            address: this._client.account.address,
            pieceCid: parsedPieceCID,
            resolvers: [
                Piece.providersResolver([this._provider]),
                ...(withCDN ? [Piece.filbeamResolver] : []),
                Piece.chainResolver,
            ],
        });
        return Piece.downloadAndValidate({
            url: pieceUrl,
            expectedPieceCid: parsedPieceCID,
        });
    }
    async getProviderInfo() {
        return await this._synapse.getProviderInfo(this.serviceProvider);
    }
    async getScheduledRemovals() {
        if (this._dataSetId == null) {
            return [];
        }
        return await PDPVerifier.getScheduledRemovals(this._client, { dataSetId: this._dataSetId });
    }
    async *getPieces(options = {}) {
        if (this._dataSetId == null) {
            return;
        }
        const batchSize = options?.batchSize ?? 100n;
        let offset = 0n;
        let hasMore = true;
        while (hasMore) {
            const result = await PDPVerifier.getActivePieces(this._client, {
                dataSetId: this._dataSetId,
                offset,
                limit: batchSize,
            });
            for (let i = 0; i < result.pieces.length; i++) {
                yield {
                    pieceCid: result.pieces[i].cid,
                    pieceId: result.pieces[i].id,
                };
            }
            hasMore = result.hasMore;
            offset += batchSize;
        }
    }
    async _getPieceIdByCID(pieceCid) {
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'getPieceIdByCID', 'Data set not found');
        }
        const parsedPieceCID = Piece.asPieceCID(pieceCid);
        if (parsedPieceCID == null) {
            throw createError('StorageContext', 'deletePiece', 'Invalid PieceCID provided');
        }
        const dataSetData = await SP.getDataSet({
            serviceURL: this._pdpEndpoint,
            dataSetId: this.dataSetId,
        });
        const pieceData = dataSetData.pieces.find((piece) => piece.pieceCid.toString() === parsedPieceCID.toString());
        if (pieceData == null) {
            throw createError('StorageContext', 'deletePiece', 'Piece not found in data set');
        }
        return pieceData.pieceId;
    }
    async deletePiece(options) {
        const { piece } = options;
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'deletePiece', 'Data set not found');
        }
        const pieceId = typeof piece === 'bigint' ? piece : await this._getPieceIdByCID(piece);
        const clientDataSetId = await this.getClientDataSetId();
        const { hash } = await schedulePieceDeletion(this._synapse.sessionClient ?? this._synapse.client, {
            serviceURL: this._pdpEndpoint,
            dataSetId: this.dataSetId,
            pieceId: pieceId,
            clientDataSetId: clientDataSetId,
        });
        return hash;
    }
    async hasPiece(options) {
        const { pieceCid } = options;
        const parsedPieceCID = Piece.asPieceCID(pieceCid);
        if (parsedPieceCID == null) {
            return false;
        }
        try {
            await SP.findPiece({
                serviceURL: this._pdpEndpoint,
                pieceCid: parsedPieceCID,
            });
            return true;
        }
        catch {
            return false;
        }
    }
    async pieceStatus(options) {
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'pieceStatus', 'Data set not found');
        }
        const parsedPieceCID = Piece.asPieceCID(options.pieceCid);
        if (parsedPieceCID == null) {
            throw createError('StorageContext', 'pieceStatus', 'Invalid PieceCID provided');
        }
        const [exists, dataSetData, currentEpoch] = await Promise.all([
            this.hasPiece({ pieceCid: parsedPieceCID }),
            SP.getDataSet({
                serviceURL: this._pdpEndpoint,
                dataSetId: this.dataSetId,
            }),
            getBlockNumber(this._client),
        ]);
        let retrievalUrl = null;
        let pieceId;
        let lastProven = null;
        let nextProofDue = null;
        let inChallengeWindow = false;
        let hoursUntilChallengeWindow = 0;
        let isProofOverdue = false;
        if (exists) {
            const [providerInfo, pdpConfig] = await Promise.all([
                this.getProviderInfo().catch(() => null),
                dataSetData != null
                    ? this._warmStorageService.getPDPConfig().catch((error) => {
                        console.debug('Failed to get PDP config:', error);
                        return null;
                    })
                    : Promise.resolve(null),
            ]);
            if (providerInfo != null) {
                retrievalUrl = createPieceUrlPDP({
                    cid: parsedPieceCID.toString(),
                    serviceURL: providerInfo.pdp.serviceURL,
                });
            }
            if (dataSetData != null && pdpConfig != null) {
                const pieceData = dataSetData.pieces.find((piece) => piece.pieceCid.toString() === parsedPieceCID.toString());
                if (pieceData != null) {
                    pieceId = pieceData.pieceId;
                    if (dataSetData.nextChallengeEpoch > 0) {
                        const challengeWindowStart = dataSetData.nextChallengeEpoch;
                        const provingDeadline = challengeWindowStart + Number(pdpConfig.challengeWindowSize);
                        nextProofDue = epochToDate(provingDeadline, this._chain.genesisTimestamp);
                        const lastProvenDate = calculateLastProofDate(dataSetData.nextChallengeEpoch, Number(pdpConfig.maxProvingPeriod), this._chain.genesisTimestamp);
                        if (lastProvenDate != null) {
                            lastProven = lastProvenDate;
                        }
                        inChallengeWindow = Number(currentEpoch) >= challengeWindowStart && Number(currentEpoch) < provingDeadline;
                        isProofOverdue = Number(currentEpoch) >= provingDeadline;
                        if (Number(currentEpoch) < challengeWindowStart) {
                            const timeUntil = timeUntilEpoch(challengeWindowStart, Number(currentEpoch));
                            hoursUntilChallengeWindow = timeUntil.hours;
                        }
                    }
                    else {
                        console.debug('Data set has nextChallengeEpoch=0, may have just been proven');
                    }
                }
            }
        }
        return {
            exists,
            dataSetLastProven: lastProven,
            dataSetNextProofDue: nextProofDue,
            retrievalUrl,
            pieceId,
            inChallengeWindow,
            hoursUntilChallengeWindow,
            isProofOverdue,
        };
    }
    async terminate() {
        if (this.dataSetId == null) {
            throw createError('StorageContext', 'terminate', 'Data set not found');
        }
        return this._synapse.storage.terminateDataSet({ dataSetId: this.dataSetId });
    }
}
