// @filoz/synapse-sdk/dist/src/storage/manager.js
import * as Piece from '@filoz/synapse-core/piece';
import { getPDPProviderByAddress } from '@filoz/synapse-core/sp-registry';
import { metadataMatches } from '@filoz/synapse-core/warm-storage';
import { zeroAddress } from 'viem';
import { CommitError, StoreError } from "../errors/storage.js";
import { SPRegistryService } from "../sp-registry/index.js";
import { combineMetadata, createError, METADATA_KEYS, SIZE_CONSTANTS, TIME_CONSTANTS } from "../utils/index.js";
import { StorageContext } from "./context.js";
const MAX_SECONDARY_ATTEMPTS = 5;
const DEFAULT_COPY_COUNT = 2;
function safeInvoke(fn, ...args) {
    if (fn == null)
        return;
    try {
        fn(...args);
    }
    catch (error) {
        console.warn('Callback error (ignored):', error instanceof Error ? error.message : error);
    }
}
export class StorageManager {
    _synapse;
    _warmStorageService;
    _withCDN;
    _defaultContexts;
    constructor(options) {
        this._synapse = options.synapse;
        this._warmStorageService = options.warmStorageService;
        this._withCDN = options.withCDN;
    }
    async upload(data, options) {
        const { contexts, explicitProviders } = await this._resolveUploadContexts(options);
        const [primary, ...secondaries] = contexts;
        let storeResult;
        try {
            storeResult = await primary.store(data, {
                pieceCid: options?.pieceCid,
                signal: options?.signal,
                onProgress: options?.callbacks?.onProgress,
            });
            safeInvoke(options?.callbacks?.onStored, primary.provider.id, storeResult.pieceCid);
        }
        catch (error) {
            throw new StoreError(`Failed to store on primary provider ${primary.provider.id} (${primary.provider.pdp.serviceURL})`, {
                cause: error instanceof Error ? error : undefined,
                providerId: primary.provider.id,
                endpoint: primary.provider.pdp.serviceURL,
            });
        }
        const pieceInputs = [{ pieceCid: storeResult.pieceCid, pieceMetadata: options?.pieceMetadata }];
        let successfulSecondaries = [];
        let pullFailures = [];
        let extraDataMap = new Map();
        if (secondaries.length > 0) {
            const pullResult = await this._pullToSecondariesWithRetry(primary, secondaries, [storeResult.pieceCid], {
                explicitProviders,
                signal: options?.signal,
                withCDN: options?.withCDN,
                metadata: options?.metadata,
                pieceMetadata: options?.pieceMetadata,
                callbacks: options?.callbacks,
                onProgress: options?.callbacks?.onPullProgress,
                onSuccess: options?.callbacks?.onCopyComplete,
                onFailure: options?.callbacks?.onCopyFailed,
                pieceInputs,
            });
            successfulSecondaries = pullResult.successful;
            pullFailures = pullResult.failures;
            extraDataMap = pullResult.extraDataMap;
        }
        const commitPromises = [
            { ctx: primary, role: 'primary' },
            ...successfulSecondaries.map((ctx) => ({ ctx, role: 'secondary' })),
        ].map(async ({ ctx, role }) => {
            const result = await ctx.commit({
                pieces: pieceInputs,
                extraData: extraDataMap.get(ctx),
                onSubmitted: (txHash) => safeInvoke(options?.callbacks?.onPiecesAdded, txHash, ctx.provider.id, [{ pieceCid: storeResult.pieceCid }]),
            });
            return { ctx, role, result };
        });
        const commitResults = await Promise.allSettled(commitPromises);
        let primaryCommit;
        let primaryCommitError;
        const secondaryCommits = [];
        const commitFailedSecondaryIds = new Set();
        for (const settled of commitResults) {
            if (settled.status === 'fulfilled') {
                const { ctx, role, result } = settled.value;
                if (role === 'primary') {
                    primaryCommit = result;
                }
                else {
                    secondaryCommits.push({ context: ctx, result });
                }
            }
            else {
                const failedIndex = commitResults.indexOf(settled);
                if (failedIndex === 0) {
                    primaryCommitError = settled.reason instanceof Error ? settled.reason : new Error(String(settled.reason));
                }
                else {
                    const failedSecondary = successfulSecondaries[failedIndex - 1];
                    commitFailedSecondaryIds.add(failedSecondary.provider.id);
                }
            }
        }
        const copies = [];
        if (primaryCommit) {
            copies.push({
                providerId: primary.provider.id,
                dataSetId: primaryCommit.dataSetId,
                pieceId: primaryCommit.pieceIds[0],
                role: 'primary',
                retrievalUrl: primary.getPieceUrl(storeResult.pieceCid),
                isNewDataSet: primaryCommit.isNewDataSet,
            });
        }
        for (const { context, result } of secondaryCommits) {
            copies.push({
                providerId: context.provider.id,
                dataSetId: result.dataSetId,
                pieceId: result.pieceIds[0],
                role: 'secondary',
                retrievalUrl: context.getPieceUrl(storeResult.pieceCid),
                isNewDataSet: result.isNewDataSet,
            });
        }
        if (copies.length === 0) {
            throw new CommitError(`Failed to commit on primary provider ${primary.provider.id} (${primary.provider.pdp.serviceURL}) - data is stored but not on-chain`, {
                cause: primaryCommitError,
                providerId: primary.provider.id,
                endpoint: primary.provider.pdp.serviceURL,
            });
        }
        if (primaryCommit) {
            safeInvoke(options?.callbacks?.onPiecesConfirmed, primaryCommit.dataSetId, primary.provider.id, [
                { pieceId: primaryCommit.pieceIds[0], pieceCid: storeResult.pieceCid },
            ]);
        }
        for (const { context, result } of secondaryCommits) {
            safeInvoke(options?.callbacks?.onPiecesConfirmed, result.dataSetId, context.provider.id, [
                { pieceId: result.pieceIds[0], pieceCid: storeResult.pieceCid },
            ]);
        }
        const failures = [...pullFailures];
        const pullFailedIds = new Set(pullFailures.map((f) => f.providerId));
        if (primaryCommitError && !pullFailedIds.has(primary.provider.id)) {
            failures.push({
                providerId: primary.provider.id,
                role: 'primary',
                error: 'Commit failed',
                explicit: explicitProviders,
            });
        }
        for (const failedId of commitFailedSecondaryIds) {
            if (!pullFailedIds.has(failedId)) {
                failures.push({
                    providerId: failedId,
                    role: 'secondary',
                    error: 'Commit failed',
                    explicit: explicitProviders,
                });
            }
        }
        return { pieceCid: storeResult.pieceCid, size: storeResult.size, copies, failures };
    }
    async _resolveUploadContexts(options) {
        if (options?.contexts != null) {
            const invalidOptions = [];
            if (options.providerIds !== undefined)
                invalidOptions.push('providerIds');
            if (options.dataSetIds !== undefined)
                invalidOptions.push('dataSetIds');
            if (options.withCDN !== undefined)
                invalidOptions.push('withCDN');
            if (invalidOptions.length > 0) {
                throw createError('StorageManager', 'upload', `Cannot specify both 'contexts' and other options: ${invalidOptions.join(', ')}`);
            }
        }
        const hasExplicitIds = (options?.providerIds != null && options.providerIds.length > 0) ||
            (options?.dataSetIds != null && options.dataSetIds.length > 0);
        const explicitProviders = options?.contexts != null || hasExplicitIds;
        const contexts = options?.contexts ??
            (await this.createContexts({
                withCDN: options?.withCDN,
                count: hasExplicitIds ? options?.count : (options?.count ?? DEFAULT_COPY_COUNT),
                metadata: options?.metadata,
                excludeProviderIds: options?.excludeProviderIds,
                providerIds: options?.providerIds,
                dataSetIds: options?.dataSetIds,
                callbacks: options?.callbacks,
            }));
        return { contexts, explicitProviders };
    }
    async _pullToSecondariesWithRetry(primary, secondaries, pieceCids, options) {
        const usedProviderIds = new Set([primary.provider.id, ...secondaries.map((s) => s.provider.id)]);
        const successful = [];
        const failures = [];
        const extraDataMap = new Map();
        for (let i = 0; i < secondaries.length; i++) {
            let currentSecondary = secondaries[i];
            let attempts = 0;
            let succeeded = false;
            while (!succeeded && attempts < MAX_SECONDARY_ATTEMPTS) {
                try {
                    let extraData;
                    if (options.pieceInputs) {
                        extraData = await currentSecondary.presignForCommit(options.pieceInputs);
                    }
                    const providerId = currentSecondary.provider.id;
                    const pullResult = await currentSecondary.pull({
                        pieces: pieceCids,
                        from: (pieceCid) => primary.getPieceUrl(pieceCid),
                        signal: options.signal,
                        extraData,
                        onProgress: options.onProgress
                            ? (cid, status) => safeInvoke(options.onProgress, providerId, cid, status)
                            : undefined,
                    });
                    if (pullResult.status === 'complete') {
                        succeeded = true;
                        successful.push(currentSecondary);
                        if (extraData) {
                            extraDataMap.set(currentSecondary, extraData);
                        }
                        for (const pieceCid of pieceCids) {
                            safeInvoke(options.onSuccess, providerId, pieceCid);
                        }
                    }
                    else {
                        const failedPieces = pullResult.pieces.filter((p) => p.status !== 'complete');
                        const errorMsg = failedPieces.length > 0
                            ? `Pull failed for ${failedPieces.length} piece(s): ${failedPieces.map((p) => p.pieceCid).join(', ')}`
                            : 'Pull failed';
                        failures.push({
                            providerId,
                            role: 'secondary',
                            error: errorMsg,
                            explicit: options.explicitProviders,
                        });
                        const err = new Error(errorMsg);
                        for (const pieceCid of pieceCids) {
                            safeInvoke(options.onFailure, providerId, pieceCid, err);
                        }
                    }
                }
                catch (error) {
                    const errorMsg = error instanceof Error ? error.message : String(error);
                    failures.push({
                        providerId: currentSecondary.provider.id,
                        role: 'secondary',
                        error: errorMsg,
                        explicit: options.explicitProviders,
                    });
                    const err = error instanceof Error ? error : new Error(errorMsg);
                    for (const pieceCid of pieceCids) {
                        safeInvoke(options.onFailure, currentSecondary.provider.id, pieceCid, err);
                    }
                }
                attempts++;
                if (!succeeded && !options.explicitProviders && attempts < MAX_SECONDARY_ATTEMPTS) {
                    try {
                        const [newContext] = await this.createContexts({
                            withCDN: options.withCDN,
                            count: 1,
                            metadata: options.metadata,
                            callbacks: options.callbacks,
                            excludeProviderIds: [...usedProviderIds],
                        });
                        currentSecondary = newContext;
                        usedProviderIds.add(newContext.provider.id);
                    }
                    catch {
                        break;
                    }
                }
                else if (!succeeded && options.explicitProviders) {
                    break;
                }
            }
        }
        return { successful, failures, extraDataMap };
    }
    async download(options) {
        if (options?.context != null) {
            const invalidOptions = [];
            if (options.providerAddress !== undefined)
                invalidOptions.push('providerAddress');
            if (options.withCDN !== undefined)
                invalidOptions.push('withCDN');
            if (invalidOptions.length > 0) {
                throw createError('StorageManager', 'download', `Cannot specify both 'context' and other options: ${invalidOptions.join(', ')}`);
            }
            return await options.context.download({
                pieceCid: options.pieceCid,
                withCDN: options.withCDN ?? this._withCDN,
            });
        }
        const parsedPieceCID = Piece.asPieceCID(options.pieceCid);
        if (parsedPieceCID == null) {
            throw createError('StorageManager', 'download', `Invalid PieceCID: ${String(options.pieceCid)}`);
        }
        const clientAddress = this._synapse.client.account.address;
        const withCDN = options.withCDN ?? this._withCDN;
        let pieceUrl;
        if (options.providerAddress) {
            const provider = await getPDPProviderByAddress(this._synapse.client, { address: options.providerAddress });
            if (provider == null) {
                throw createError('StorageManager', 'download', `Provider ${options.providerAddress} not found`);
            }
            pieceUrl = Piece.createPieceUrlPDP({ cid: parsedPieceCID.toString(), serviceURL: provider.pdp.serviceURL });
        }
        else {
            try {
                pieceUrl = await Piece.resolvePieceUrl({
                    client: this._synapse.client,
                    address: clientAddress,
                    pieceCid: parsedPieceCID,
                    resolvers: [
                        ...(withCDN ? [Piece.filbeamResolver] : []),
                        Piece.chainResolver,
                        Piece.providersResolver(this._defaultContexts?.map((context) => context.provider) ?? []),
                    ],
                });
            }
            catch (error) {
                throw createError('StorageManager', 'download', `All provider retrieval attempts failed and no additional retriever method was configured`, error);
            }
        }
        return Piece.downloadAndValidate({
            expectedPieceCid: parsedPieceCID,
            url: pieceUrl,
        });
    }
    async preflightUpload(options) {
        let withCDN = options?.withCDN ?? this._withCDN;
        if (options?.metadata != null && METADATA_KEYS.WITH_CDN in options.metadata) {
            const value = options.metadata[METADATA_KEYS.WITH_CDN];
            if (value !== '') {
                console.warn(`Warning: withCDN metadata entry has unexpected value "${value}". Expected empty string.`);
            }
            withCDN = true;
        }
        return await StorageContext.performPreflightCheck({
            warmStorageService: this._warmStorageService,
            size: options.size,
            withCDN,
        });
    }
    async createContexts(options) {
        const withCDN = options?.withCDN ?? this._withCDN;
        const canUseDefault = options == null || (options.providerIds == null && options.dataSetIds == null);
        if (this._defaultContexts != null) {
            const expectedSize = options?.count ?? DEFAULT_COPY_COUNT;
            if (this._defaultContexts.length === expectedSize &&
                this._defaultContexts.every((context) => options?.excludeProviderIds?.includes(context.provider.id) !== true)) {
                const requestedMetadata = combineMetadata(options?.metadata, withCDN);
                if (this._defaultContexts.every((defaultContext) => metadataMatches(defaultContext.dataSetMetadata, requestedMetadata))) {
                    if (options?.callbacks != null) {
                        for (const defaultContext of this._defaultContexts) {
                            try {
                                options.callbacks.onProviderSelected?.(defaultContext.provider);
                            }
                            catch (error) {
                                console.error('Error in onProviderSelected callback:', error);
                            }
                            if (defaultContext.dataSetId != null) {
                                try {
                                    options.callbacks.onDataSetResolved?.({
                                        dataSetId: defaultContext.dataSetId,
                                        provider: defaultContext.provider,
                                    });
                                }
                                catch (error) {
                                    console.error('Error in onDataSetResolved callback:', error);
                                }
                            }
                        }
                    }
                    return this._defaultContexts;
                }
            }
        }
        const contexts = await StorageContext.createContexts({
            synapse: this._synapse,
            warmStorageService: this._warmStorageService,
            ...options,
            withCDN,
        });
        if (canUseDefault) {
            this._defaultContexts = contexts;
        }
        return contexts;
    }
    async createContext(options) {
        const effectiveWithCDN = options?.withCDN ?? this._withCDN;
        const canUseDefault = options == null || (options.providerId == null && options.dataSetId == null);
        if (canUseDefault && this._defaultContexts != null) {
            const requestedMetadata = combineMetadata(options?.metadata, effectiveWithCDN);
            for (const defaultContext of this._defaultContexts) {
                if (options?.excludeProviderIds?.includes(defaultContext.provider.id)) {
                    continue;
                }
                if (!metadataMatches(defaultContext.dataSetMetadata, requestedMetadata)) {
                    continue;
                }
                if (options?.callbacks != null) {
                    try {
                        options.callbacks.onProviderSelected?.(defaultContext.provider);
                    }
                    catch (error) {
                        console.error('Error in onProviderSelected callback:', error);
                    }
                    if (defaultContext.dataSetId != null) {
                        try {
                            options.callbacks.onDataSetResolved?.({
                                dataSetId: defaultContext.dataSetId,
                                provider: defaultContext.provider,
                            });
                        }
                        catch (error) {
                            console.error('Error in onDataSetResolved callback:', error);
                        }
                    }
                }
                return defaultContext;
            }
        }
        const context = await StorageContext.create({
            synapse: this._synapse,
            warmStorageService: this._warmStorageService,
            ...options,
            withCDN: effectiveWithCDN,
        });
        if (canUseDefault) {
            this._defaultContexts = [context];
        }
        return context;
    }
    async getDefaultContext() {
        return await this.createContext();
    }
    async findDataSets(options = {}) {
        const { address = this._synapse.client.account.address } = options;
        return await this._warmStorageService.getClientDataSetsWithDetails({ address });
    }
    async terminateDataSet(options) {
        return this._warmStorageService.terminateDataSet(options);
    }
    async getStorageInfo() {
        const chain = this._synapse.client.chain;
        try {
            const getOptionalAllowances = async () => {
                try {
                    const approval = await this._synapse.payments.serviceApproval();
                    return {
                        service: chain.contracts.fwss.address,
                        isApproved: approval.isApproved,
                        rateAllowance: approval.rateAllowance,
                        lockupAllowance: approval.lockupAllowance,
                        rateUsed: approval.rateUsage,
                        lockupUsed: approval.lockupUsage,
                    };
                }
                catch {
                    return null;
                }
            };
            const spRegistry = new SPRegistryService({ client: this._synapse.client });
            const [pricingData, approvedIds, allowances] = await Promise.all([
                this._warmStorageService.getServicePrice(),
                this._warmStorageService.getApprovedProviderIds(),
                getOptionalAllowances(),
            ]);
            const providers = await spRegistry.getProviders({ providerIds: approvedIds });
            const epochsPerMonth = BigInt(pricingData.epochsPerMonth);
            const noCDNPerEpoch = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / epochsPerMonth;
            const withCDNPerEpoch = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / epochsPerMonth;
            const noCDNPerDay = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / TIME_CONSTANTS.DAYS_PER_MONTH;
            const withCDNPerDay = BigInt(pricingData.pricePerTiBPerMonthNoCDN) / TIME_CONSTANTS.DAYS_PER_MONTH;
            const validProviders = providers.filter((p) => p.serviceProvider !== zeroAddress);
            return {
                pricing: {
                    noCDN: {
                        perTiBPerMonth: BigInt(pricingData.pricePerTiBPerMonthNoCDN),
                        perTiBPerDay: noCDNPerDay,
                        perTiBPerEpoch: noCDNPerEpoch,
                    },
                    withCDN: {
                        perTiBPerMonth: BigInt(pricingData.pricePerTiBPerMonthNoCDN),
                        perTiBPerDay: withCDNPerDay,
                        perTiBPerEpoch: withCDNPerEpoch,
                    },
                    tokenAddress: pricingData.tokenAddress,
                    tokenSymbol: 'USDFC',
                },
                providers: validProviders,
                serviceParameters: {
                    epochsPerMonth,
                    epochsPerDay: TIME_CONSTANTS.EPOCHS_PER_DAY,
                    epochDuration: TIME_CONSTANTS.EPOCH_DURATION,
                    minUploadSize: SIZE_CONSTANTS.MIN_UPLOAD_SIZE,
                    maxUploadSize: SIZE_CONSTANTS.MAX_UPLOAD_SIZE,
                },
                allowances,
            };
        }
        catch (error) {
            throw new Error(`Failed to get storage service information: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
}
