// @filoz/synapse-sdk/dist/src/errors/storage.js
import { isSynapseError, SynapseError } from '@filoz/synapse-core/errors';
export class StoreError extends SynapseError {
    name = 'StoreError';
    providerId;
    endpoint;
    constructor(message, options) {
        super(message, options);
        this.providerId = options?.providerId?.toString();
        this.endpoint = options?.endpoint;
    }
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            providerId: this.providerId,
            endpoint: this.endpoint,
        };
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'StoreError';
    }
}
export class CommitError extends SynapseError {
    name = 'CommitError';
    providerId;
    endpoint;
    constructor(message, options) {
        super(message, options);
        this.providerId = options?.providerId?.toString();
        this.endpoint = options?.endpoint;
    }
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            providerId: this.providerId,
            endpoint: this.endpoint,
        };
    }
    static is(value) {
        return isSynapseError(value) && value.name === 'CommitError';
    }
}
