// @filoz/synapse-core/dist/src/sp/pull.js
import { asChain } from "../chains.js";
import { signAddPieces } from "../typed-data/sign-add-pieces.js";
import { signCreateDataSetAndAddPieces } from "../typed-data/sign-create-dataset-add-pieces.js";
import { datasetMetadataObjectToEntry, pieceMetadataObjectToEntry } from "../utils/metadata.js";
import { randU256 } from "../utils/rand.js";
import * as PDP from "./sp.js";
function isExistingDataSet(options) {
    return options.dataSetId != null && options.dataSetId > 0n;
}
function toSigningPieces(pieces) {
    return pieces.map((piece) => ({
        pieceCid: piece.pieceCid,
        metadata: pieceMetadataObjectToEntry(piece.metadata),
    }));
}
function toPullPieces(pieces) {
    return pieces.map((piece) => ({
        pieceCid: piece.pieceCid.toString(),
        sourceUrl: piece.sourceUrl,
    }));
}
async function signPullExtraData(client, options) {
    if (isExistingDataSet(options)) {
        return signAddPieces(client, {
            clientDataSetId: options.clientDataSetId,
            nonce: options.nonce,
            pieces: toSigningPieces(options.pieces),
        });
    }
    return signCreateDataSetAndAddPieces(client, {
        clientDataSetId: options.clientDataSetId ?? randU256(),
        payee: options.payee,
        payer: options.payer,
        metadata: datasetMetadataObjectToEntry(options.metadata, {
            cdn: options.cdn ?? false,
        }),
        nonce: options.nonce,
        pieces: toSigningPieces(options.pieces),
    });
}
async function resolvePullParams(client, options) {
    const chain = asChain(client.chain);
    return {
        serviceURL: options.serviceURL,
        recordKeeper: options.recordKeeper ?? chain.contracts.fwss.address,
        extraData: options.extraData ?? (await signPullExtraData(client, options)),
        dataSetId: isExistingDataSet(options) ? options.dataSetId : undefined,
        pieces: toPullPieces(options.pieces),
        signal: options.signal,
    };
}
export async function pullPieces(client, options) {
    return PDP.pullPieces(await resolvePullParams(client, options));
}
export async function waitForPullStatus(client, options) {
    const params = await resolvePullParams(client, options);
    return PDP.waitForPullStatus({
        ...params,
        onStatus: options.onStatus,
        timeout: options.timeout,
        pollInterval: options.pollInterval,
    });
}
